azhidkov.pro/content/posts/21/04/en_why-jpa-should-be-avoided.adoc

```diff
@@ -0,0 +1,359 @@
+
+
+JPA undermines both design and performance.
+First, let’s take a look at how JPA can jeopardize software design.
+
+=== Default constructors
+
+JPA requires that all entity classes include constructors by default:
+[quote, JSR 338: JavaTM Persistence API; Version 2.2; "2.1 The Entity Class," https://github.com/javaee/jpa-spec/blob/master/jsr338-MR/JavaPersistence.pdf]
+____
+The entity class must have a no-arg constructor.
+____
+
+Note that default constructors are an antipattern and a ticking bomb, as they allow for invariant-violating objects. They also cause link:++{{<ref "/posts/21/01/210119-effects#_временная_связность">}}++[temporal coupling].
+Look https://blog.ploeh.dk/2011/05/30/DesignSmellDefaultConstructor/[here] for more details.
+
+You can partly avoid this problem by making the default constructor package private and marking it as @Deprecated.
+
+I've never seen anyone but me adhere to this practice, though.
+
+=== Classes must be open for inheritance
+
+JPA requires entity classes to be open for inheritance:
+[quote, JSR 338: JavaTM Persistence API; Version 2.2; "2.1 The Entity Class"]
+____
+The entity class must not be final
+____
+However, you must either design and document your classes for inheritance or prohibit it.
+I'll quote the classics here: https://www.amazon.com/Effective-Java-Joshua-Bloch/dp/0134685997[Effective Java], chapter "Item 19:  Design and document for inheritance or else prohibit it."
+
+It would take much more effort to design a class for inheritance than to define a data structure with a bunch of fields, as well as setters and getters for these.
+
+I've never seen a JPA Entity designed with inheritance in mind.
+
+Although JPA's entity inheritability can cause potential problems, I've never encountered any in my experience.
+
+=== Objects must be mutable
+
+JPA can't work with immutable objects "By Design," and mutability is inherent to its specification:
+[quote, JSR 338: JavaTM Persistence API; Version 2.2; "3.2.4 Synchronization to the Database"]
+____
+An update to the state of an entity includes both the assignment of a new value to a persistent property or field of the entity as well as the modification of a mutable value of a persistent property or field
+____
+If your entire model is mutable, though, you get all the issues with:
+
+. link:++{{<ref "/posts/21/01/210119-effects#_временная_связность">}}++[temporal cohesion];
+. link:++{{<ref "/posts/21/01/210119-effects#_нелокальность_рассуждений">}}++[locality of reference];
+. link:++{{<ref "/posts/21/01/210119-effects#_конкурентное_программирование">}}++[concurrent development].
+
+To minimize its abstraction https://en.wikipedia.org/wiki/Leaky_abstraction[leaks], JPA needs to make sure an object in memory only corresponds to just one row in the table.
+So if instead of mutating an object you create a new instance of it with an updated state, JPA will treat it as a new object. Naturally, it'll also connect it to a new table row.
```

Comment:
```suggestion
So if instead of mutating an object you create a new instance of it with an updated state, JPA will treat it as a new object. Naturally, it'll also link/relate it to a new table row.
```